### PR TITLE
Implement a stub server printer for CRDs

### PR DESCRIPTION
**Pull request by smarterclayton**
This wires up TableConvertor to CRDs and puts a basic implementation in place for custom paths. However, since our OpenAPISchema can't store OpenAPI extension fields there is no way to expose the custom column piece that get.go supports today (`x-kubernetes-print-columns`). That piece can be implemented separately and needs discussion.

As this is purely exposing the default interface, very low risk. Will add an e2e test that covers this under a registered CRD.

@soltysh @sttts @kubernetes/sig-api-machinery-pr-reviews

A couple of options for wiring up the actual definition:

1. add a new "extensions" map to spec.validation
   1. Downside: won't handle future child nested fields, not the correct schema
2. try to change the OpenAPISchema3 field to support extensions
   1. Would require a breaking protobuf change, is also very difficult
   2. Could store the entire schema as opaque JSON and then parse on load (might be the right thing anyway)
3. Support this as an annotation in 1.11 - `alpha.customresource.k8s.io/x-kubernetes-print-columns` like the CLI

Part of #58536

### CODE REVIEW

**Comment by smarterclayton** [comment]
Given how late we are, and the limitations in existing CRD, i think we'll need to punt custom columns to 1.11.  However, this PR enables the default columns so I'd like to get it in (there's no complexity in the core path)

**Comment by smarterclayton** [comment]
@sttts @soltysh @kubernetes/sig-api-machinery-pr-reviews equivalent to #60219

**Review by soltysh — APPROVED** [review]
lgtm

**Comment by jberkus** [comment]
@smarterclayton if this is staying in 1.10, please add kind/ priority/ and status/approved-for-milestone labels to it.  Otherwise it will be automatically kicked out on monday.  Thanks!

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -447,6 +448,11 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
 		scaleSpec = crd.Spec.Subresources.Scale
 	}
 
+	table, err := tableconvertor.New(nil)
+	if err != nil {
+		glog.V(2).Infof("The CRD for %v has an invalid printer specification, falling back to default printing: %v", schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version}, err)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: the output misses the CRD kind.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -447,6 +448,11 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
 		scaleSpec = crd.Spec.Subresources.Scale
 	}
 
+	table, err := tableconvertor.New(nil)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: // TODO: replace the nil with the actual extensions when we support them in OpenAPI validation specs.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go [hunk 3]
```diff
@@ -71,6 +72,8 @@ func newStorage(t *testing.T) (customresource.CustomResourceStorage, *etcdtestin
 
 	status := &apiextensions.CustomResourceSubresourceStatus{}
 
+	table, _ := tableconvertor.New(nil)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go]: // TODO: .... as above.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	var err error
+	buf := &bytes.Buffer{}
+	table.Rows, err = metatable.MetaToTableRow(obj, func(obj runtime.Object, m metav1.Object, name, age string) ([]interface{}, error) {
+		cells := make([]interface{}, 2, 2+len(c.columns))
+		cells[0] = name
+		cells[1] = age
+		for _, column := range c.columns {
+			if err := column.Execute(buf, obj); err != nil {
+				cells = append(cells, nil)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go]: should be distinguish between not found paths and unprintable data?

Comment (smarterclayton) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go]: Possibly.  I think of the current annotation from OpenAPI is underspecified and a bit of a quick hack, so some more thought is deserved for all of this.  I was going to propose a 1.11 design extension that covered more of the edge cases - I think given server side printing depends on the CRD, not just OpenAPI, we can do a better job of making this useful.  The design goal of CRD is to invert the API model and make clients able to usefully function against a dumb api server, so a better definition (server side smarts) seems logical.

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by smarterclayton** [comment]
/retest

might be related, although the stack output doesn't help

**Comment by smarterclayton** [comment]
/retest

weird, passes locally

**Comment by sttts** [comment]
From Slack:

Stefan Schimanski [18:16] @claytonc which kinds of fields do we have to support? I guess simple json paths .word.word style are enough.
claytonc [18:16]i think anything more complex than that will have problems
we might want to gather feedback
from actual CRDs
like tectonic

Stefan Schimanski [18:17] @claytonc what I don't like about openapi x-kubectl-column style extensions is that with quantors this becomes ambiguous easily
`allOf: [someSchemaWithColumns, anotherSchemaWithColumns]`
a flat json path list spec.validation sounds more reasonable

claytonc [18:18] i don’t like x-kubectl-columns for a lot of reasons
but yes

**Comment by sttts** [comment]
@brancz can you comment on your requirements for columns?

**Comment by brancz** [comment]
For us simple json paths would definitely suffice.

**Pull request merged**